source/opt/scalar_replacement_pass.cpp

```diff
@@ -0,0 +1,598 @@
+
+  uint32_t initId = source->GetSingleWordInOperand(1u);
+  uint32_t storageId = GetStorageType(newVar)->result_id();
+  ir::Instruction* init = get_def_use_mgr()->GetDef(initId);
+  uint32_t newInitId = 0;
+  if (init->opcode() == SpvOpConstantNull) {
+    // Initialize to appropriate NULL.
+    auto iter = type_to_null_.find(storageId);
+    if (iter == type_to_null_.end()) {
+      newInitId = TakeNextId();
+      type_to_null_[storageId] = newInitId;
+      context()->AddGlobalValue(MakeUnique<ir::Instruction>(
+          context(), SpvOpConstantNull, storageId, newInitId,
+          std::initializer_list<ir::Operand>{}));
+      ir::Instruction* newNull = &*--context()->types_values_end();
+      get_def_use_mgr()->AnalyzeInstDefUse(newNull);
+    } else {
+      newInitId = iter->second;
+    }
+  } else if (ir::IsSpecConstantInst(init->opcode())) {
+    // Generate a constant for the extraction.
+    uint32_t constantId = 0;
+    auto iter = index_to_constant_.find(index);
+    if (iter == index_to_constant_.end()) {
+      if (int_id_ == 0) {
+        int_id_ = GetIntId();
+      }
+      constantId = TakeNextId();
+      context()->AddGlobalValue(MakeUnique<ir::Instruction>(
+          context(), SpvOpConstant, int_id_, constantId,
+          std::initializer_list<ir::Operand>{
+              {SPV_OPERAND_TYPE_LITERAL_INTEGER, {index}}}));
+      ir::Instruction* constant = &*--context()->types_values_end();
+      get_def_use_mgr()->AnalyzeInstDefUse(constant);
+    } else {
+      constantId = iter->second;
+    }
+
+    // Create a new constant extract.
+    newInitId = TakeNextId();
+    context()->AddGlobalValue(MakeUnique<ir::Instruction>(
+        context(), SpvOpSpecConstantOp, storageId, newInitId,
+        std::initializer_list<ir::Operand>{
+            {SPV_OPERAND_TYPE_SPEC_CONSTANT_OP_NUMBER, {SpvOpCompositeExtract}},
+            {SPV_OPERAND_TYPE_ID, {constantId}}}));
+    ir::Instruction* newSpecConst = &*--context()->types_values_end();
+    get_def_use_mgr()->AnalyzeInstDefUse(newSpecConst);
+  } else if (init->opcode() == SpvOpConstantComposite) {
+    // Get the appropriate index constant.
+    newInitId = init->GetSingleWordInOperand(index);
+    ir::Instruction* element = get_def_use_mgr()->GetDef(newInitId);
+    if (element->opcode() == SpvOpUndef) {
+      // Undef is not a valid initializer for a variable.
+      newInitId = 0;
+    }
+  } else {
+    assert(false);
+  }
+
+  if (newInitId != 0) {
+    newVar->AddOperand({SPV_OPERAND_TYPE_ID, {newInitId}});
+  }
+}
+
+size_t ScalarReplacementPass::GetIntegerLiteral(const ir::Operand& op) const {
+  assert(op.words.size() <= 2);
+  size_t len = 0;
+  for (uint32_t i = 0; i != op.words.size(); ++i) {
+    len |= (op.words[i] << (32 * i));
+  }
+  return len;
+}
+
+size_t ScalarReplacementPass::GetConstantInteger(
+    const ir::Instruction* constant) const {
+  assert(constant->opcode() == SpvOpConstant ||
+         constant->opcode() == SpvOpConstantNull);
```

Comment:
Should assert in the type as well?  OpConstant is uses for floating point constants as well.

Reply:
Sure I can add that.